### PR TITLE
feat: add cluster client

### PR DESCRIPTION
**Pull request by LingKa28**
Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
Add cluster client.

* what changes does this pull request make?
  * add cluster client.

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
No.

### CODE REVIEW

File: client/protocol.go [hunk 1]
```diff
@@ -129,10 +129,29 @@ func fastFetchCluster(addrs []string, proposeTimeout time.Duration) (*curpapi.Fe
...
+		if err != nil {
+			cluster, err := c.fetchCluster(false)
+			if err != nil {
+				return nil, err
+			}
+			err = c.setCluster(cluster)
+			if err != nil {
+				return nil, err
+			}
+			time.Sleep(2 * time.Second)
```
Comment (bsbds) [inline client/protocol.go:150]: There's no need to sleep here.

File: client/protocol.go [hunk 2]
```diff
@@ -424,23 +431,43 @@ OuterLoop:
...
+				break
 			}
-		case err := <-errCh:
-			c.logger.Warn("fetch cluster error", zap.Error(err))
 		}
+
+		c.logger.Info("fetch cluster succeeded")
+		c.state.leader = *res.LeaderId
+		c.state.term = res.Term
+		c.state.checkAndUpdate(*res.LeaderId, res.Term)
```
Comment (bsbds) [inline client/protocol.go:442]: We should return here if the fetch is successful, otherwise it will continue to retry.

File: client/protocol.go [hunk 3]
```diff
@@ -505,6 +532,45 @@ type state struct {
...
+			if s.leader == newLeaderID {
+				panic("there should never be two leader in one term")
+			}
+		}
+	}
+}
+
+// Set the leader and notify all the waiters
+func (s *state) setLeader(id ServerId) {
+	fmt.Println("client update its leader to", id)
```
Comment (bsbds) [inline client/protocol.go:561]: Please remove this print

File: client/cluster.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+func (c *clusterClient) MemberAddAsLearner(ctx context.Context, peerAddrs []string) (*MemberAddResponse, error) {
+	return c.memberAdd(ctx, peerAddrs, true)
+}
+
+func (c *clusterClient) memberAdd(ctx context.Context, peerAddrs []string, isLeader bool) (*MemberAddResponse, error) {
+	req := &xlineapi.MemberAddRequest{
+		PeerURLs:  peerAddrs,
+		IsLearner: isLeader,
```
Comment (themanforfree) [inline client/cluster.go:74]: `IsLearner`, not `IsLeader`

File: client/protocol.go [hunk 5]
```diff
@@ -129,10 +129,29 @@ func fastFetchCluster(addrs []string, proposeTimeout time.Duration) (*curpapi.Fe
...
+			if err != nil {
+				return nil, err
+			}
+			err = c.setCluster(cluster)
+			if err != nil {
+				return nil, err
+			}
+			time.Sleep(2 * time.Second)
+			continue
+		}
```
Comment (themanforfree) [inline client/protocol.go:152]: Please check err type, not every err needs a fetch cluster.

File: client/cluster.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (LingKa28) [reply client/cluster.go:74]: Done.

File: client/protocol.go [hunk 6]
```diff
@@ -237,6 +256,7 @@ func (c *protocolClient) fastRound(cmd *xlineapi.Command) (*fastRoundRes, error)
 	for i := 0; i < len(c.connects); i++ {
 		select {
 		case res := <-resCh:
+			// c.state.checkAndUpdate(*res.LeaderId, res.Term)
```
Comment (themanforfree) [inline client/protocol.go:259]: If it's useless, remove it

File: client/protocol.go [hunk 7]
```diff
@@ -379,38 +399,25 @@ func (c *protocolClient) newSeqNum() uint64 {
...
+		retryTimeout := timeout.nextRetry()
+		for _, conn := range c.connects {
+			protocolClient := curpapi.NewProtocolClient(conn)
+			ctx, cancel := context.WithTimeout(context.Background(), retryTimeout)
+			defer cancel()
+			r, err := protocolClient.FetchCluster(ctx, &curpapi.FetchClusterRequest{Linearizable: linearizable})
+			if err != nil {
+				c.logger.Warn("fetch cluster error", zap.Error(err))
+				continue
+			}
```
Comment (themanforfree) [inline client/protocol.go:420]: This refactoring will cause `FetchCluster` to fail to be concurrent, slowing down the request speed

File: client/cluster.go (hunk 4 above)
Comment (LingKa28) [reply client/cluster.go:74]: Done.

File: client/protocol.go (hunk 1 above)
Comment (LingKa28) [reply client/protocol.go:150]: Done.

File: client/protocol.go (hunk 2 above)
Comment (LingKa28) [reply client/protocol.go:442]: Done.

File: client/protocol.go (hunk 3 above)
Comment (LingKa28) [reply client/protocol.go:561]: Done.

File: client/protocol.go (hunk 5 above)
Comment (LingKa28) [reply client/protocol.go:152]: Done.

File: tests/cluster_test.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+		res, _ := clusterClient.MemberUpdate(context.TODO(), id1, []string{addr3})
+		for _, m := range res.Members {
+			if m.ID == id1 {
+				assert.Equal(t, addr3, m.PeerURLs[0])
+			}
+		}
+	})
+
+	t.Run("cluster_promote_member", func(t *testing.T) {
+		res, _ := clusterClient.MemberPromote(context.TODO(), id1)
```
Comment (Phoenix500526) [inline tests/cluster_test.go:46]: Why promote id1 and the id2 will become a learner?

File: client/protocol.go (hunk 6 above)
Comment (LingKa28) [reply client/protocol.go:259]: Done.

File: tests/cluster_test.go (hunk 8 above)
Comment (LingKa28) [reply tests/cluster_test.go]: I misunderstood the semantics of promote, and I've changed it to the correct logic.

File: client/protocol.go (hunk 7 above)
Comment (LingKa28) [reply client/protocol.go]: Thanks for the suggestion, I've changed it to be concurrent.

File: client/protocol.go [hunk 9]
```diff
@@ -129,10 +130,28 @@ func fastFetchCluster(addrs []string, proposeTimeout time.Duration) (*curpapi.Fe
...
+	var res *proposeRes
+	var err error
+	for {
+		if useFastPath {
+			res, err = c.fastPath(cmd)
+		} else {
+			res, err = c.slowPath(cmd)
+		}
+
+		if errors.Is(err, errors.New("wrong cluster version")) {
```
Comment (iGxnon) [inline client/protocol.go:142]: make a const `wrong cluster version` error might be better

Comment (LingKa28) [reply client/protocol.go]: You have right, and I've made the changes you suggested.

**Review by iGxnon — APPROVED** [review]
LGTM

**Review by bsbds — APPROVED** [review]

**Pull request merged**